### PR TITLE
[SCP-2937] Upgrade to PureScript 0.14.5

### PR DESCRIPTION
**Pull request by jhbertra**
To help parse the significant number of changes in this PR, here is the high-level list:

- Upgraded to PS 14 (includes breaking change fixes and Halogen 6 upgrade)
- Removed local web-common and switched to separate repo
- Update PSGenerator for updated purescript-bridge
- Switched from foreign-generic to argonaut for JSON handling on the frontend
- Switched from `BigInteger` type defined in forked foreign-generic to `Data.BigInt.Argonaut` defined in web-common.

I suggest not getting caught in the details of the PR and focusing mainly on testing the playground, as it is quite likely that there are some regressions caused by such a large change.

I have tested everything fairly thoroughly, and done a systematic sweep for the most likely bugs, but additional QA testing would be appreciated.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: marlowe-playground-client/src/Marlowe/Parser.js [hunk 1]
```diff
@@ -18,6 +18,8 @@ exports.parse_ = function (emptyInputError, parserError, success, fs, input) {
           return success(parser.results[0]);
         }
     } catch (error) {
+        console.log(input);
+        console.log(message);
```
Comment (palas) [inline marlowe-playground-client/src/Marlowe/Parser.js:22]: This looks like debug code

**Review by palas — APPROVED** [review]
Looks good! Nice simplifications

File: default.nix [hunk 2]
```diff
@@ -38,13 +38,9 @@ rec {
 
   # TODO This stuff should probably be exposed as an overlay in the plutus-apps if
   # we switch to flakes.
-  webCommon = pkgs.callPackage (sources.plutus-apps + "/web-common") { inherit (marlowe.lib) gitignore-nix; };
+  webCommon = pkgs.callPackage sources.web-common { inherit (marlowe.lib) gitignore-nix; };
```
Comment (hrajchert) [inline default.nix:41]: where is the web-common coming from now? How can I override it with a local implementation?

File: marlowe-dashboard-client/src/Capability/PlutusApps/MarloweApp/Types.purs [hunk 3]
```diff
@@ -32,11 +40,24 @@ data LastResult
 
 derive instance genericLastResult :: Generic LastResult _
 
-instance encodeLastResult :: Encode LastResult where
-  encode a = genericEncode defaultOptions a
-
-instance decodeLastResult :: Decode LastResult where
-  decode a = genericDecode defaultOptions a
+instance encodeLastResult :: EncodeJson LastResult where
```
Comment (hrajchert) [inline marlowe-dashboard-client/src/Capability/PlutusApps/MarloweApp/Types.purs:43]: Why did you change the way this was encoded? is genericEncode no longer an option?

**Review by hrajchert — APPROVED** [review]
LGTM

**Review by yveshauser — APPROVED** [review]
:+1:

File: default.nix (hunk 2 above)
Comment (jhbertra) [reply default.nix:41]: `web-common` is now its own repo - https://github.com/input-output-hk/purescript-web-common.

It exposes a standard package set that defines its self, from which we inherit in the project `packages.dhall`. You can override the PureScript dependencies locally like this:

```bash
# clone the web-common repo
$ git clone git@github.com:input-output-hk/purescript-web-common.git
$ cd purescript-web-common

# make local changes ...

# open the packages.dhall file in marlowe-cardano
$ cd ../marlowe-cardano
$ open ./packages.dhall
```

Override `web-common`:

```
let upstream =
      https://github.com/input-output-hk/purescript-web-common/releases/download/v1.1.4/packages.dhall sha256:ab905184ee034c35867e73ace0475ce29d1a0b975581d355af6f35e680449749

let overrides = upstream
  with web-common = ../purescript-web-common/spago.dhall as Location

let additions = {=}

in  upstream // overrides // additions
```

As for using a local version for the `static` files, for now you can update the entry in `shell.nix` that says `export WEB_COMMON_SRC="${webCommon}"` to `export WEB_COMMON_SRC=<absolute-path-to-web-common-repo>` And it should work with `npm run build:webpack:dev` /  `npm start`.

I am in the process of creating an `npm` module for `web-common`, after which you should be able to use `npm link` to manage this.

File: marlowe-dashboard-client/src/Capability/PlutusApps/MarloweApp/Types.purs (hunk 3 above)
Comment (jhbertra) [reply marlowe-dashboard-client/src/Capability/PlutusApps/MarloweApp/Types.purs:43]: No, it isn't - that was from our patched version of `foreign-generic`. There is a `argonaut-generic` that works in a similar way, but the encoding is not compatible with Aeson. Hence the  explicit encode / decode instances.
